### PR TITLE
[feature-wip](resource-queue) create resource queue to submit async jobs

### PR DESCRIPTION
**Pull request by AshinGau**
# Proposed changes

DSIP-025: https://cwiki.apache.org/confluence/display/DORIS/DSIP-025%3A+Support+Resource+Queue

## Problem summary
Support create resource queue to submit async jobs.

**Create Resource Queue**
```
mysql> create resource_queue test_queue1 with resource () with matching policy ("user"="ro*", "ip"="127.0.0.*");

mysql> show all resource_queue;
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+
| QueueId | QueueName   | PendingJobs | RunningJobs | QueueConfig                                  | MatchingPolicy                 |
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+
|   74023 | test_queue1 |           0 |           0 | "max_concurrency"="1", "max_queue_size"="10" | "user"="ro*", "ip"="127.0.0.*" |
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+

```

**Submit SQL Job with Matching Policy**
```
mysql> submit sql job (insert into internal.test.customer select * from hive.tpch1.customer);
Query OK, 0 rows affected (32.22 sec)
{'label':'insert_b1a5d9d2ab9240d9_b5389e59e7799e21', 'status':'PENDING', 'jobId':'job_74024', 'resource_queue':'test_queue1'}

mysql> show all resource_queue;
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+
| QueueId | QueueName   | PendingJobs | RunningJobs | QueueConfig                                  | MatchingPolicy                 |
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+
|   74023 | test_queue1 |           0 |           1 | "max_concurrency"="1", "max_queue_size"="10" | "user"="ro*", "ip"="127.0.0.*" |
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+
```

**Submit SQL Job with Specified Queue Name**
```
mysql> submit sql job (insert into internal.test.customer select * from hive.tpch1.customer) properties ("resource_queue"="test_queue2");
Query OK, 0 rows affected (0.00 sec)
{'label':'insert_a7a9f83bb5f6426b_ab62912378f07d6d', 'status':'PENDING', 'jobId':'job_74027', 'resource_queue':'test_queue2'}

mysql> show all resource_queue;
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+
| QueueId | QueueName   | PendingJobs | RunningJobs | QueueConfig                                  | MatchingPolicy                 |
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+
|   74023 | test_queue1 |           0 |           1 | "max_concurrency"="1", "max_queue_size"="10" | "user"="ro*", "ip"="127.0.0.*" |
|   74026 | test_queue2 |           0 |           1 | "max_concurrency"="1", "max_queue_size"="0"  | "user"="root", "ip"="*"        |
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+
```

**Find the Best Queue**
```
mysql> create resource_queue test_queue3 with resource ("max_concurrency"="2", "max_queue_size"="5") with matching policy ();
Query OK, 0 rows affected (0.00 sec)

mysql> show all resource_queue;
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+
| QueueId | QueueName   | PendingJobs | RunningJobs | QueueConfig                                  | MatchingPolicy                 |
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+
|   74023 | test_queue1 |           2 |           1 | "max_concurrency"="1", "max_queue_size"="10" | "user"="ro*", "ip"="127.0.0.*" |
|   74026 | test_queue2 |           0 |           1 | "max_concurrency"="1", "max_queue_size"="0"  | "user"="root", "ip"="*"        |
|   74031 | test_queue3 |           0 |           0 | "max_concurrency"="2", "max_queue_size"="5"  | "user"="root", "ip"="*"        |
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+

mysql> submit sql job (insert into internal.test.customer select * from hive.tpch1.customer);
Query OK, 0 rows affected (0.01 sec)
{'label':'insert_2ef98ee415ae4c6c_95976dfebebadf36', 'status':'PENDING', 'jobId':'job_74032', 'resource_queue':'test_queue3'}
```

**Alter Resource Queue**
```
mysql> alter resource_queue test_queue1 set resource () set matching policy ();
ERROR 1105 (HY000): errCode = 2, detailMessage = Should alter resource queue when empty(numPendingJobs=2, numRunningJobs=1)
mysql> create resource_queue test_queue4 with resource ("max_concurrency"="2", "max_queue_size"="5") with matching policy ();

mysql> show all resource_queue;
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+
| QueueId | QueueName   | PendingJobs | RunningJobs | QueueConfig                                  | MatchingPolicy                 |
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+
|   74033 | test_queue4 |           0 |           0 | "max_concurrency"="2", "max_queue_size"="5"  | "user"="root", "ip"="*"        |
|   74023 | test_queue1 |           2 |           1 | "max_concurrency"="1", "max_queue_size"="10" | "user"="ro*", "ip"="127.0.0.*" |
|   74026 | test_queue2 |           0 |           1 | "max_concurrency"="1", "max_queue_size"="0"  | "user"="root", "ip"="*"        |
|   74031 | test_queue3 |           0 |           1 | "max_concurrency"="2", "max_queue_size"="5"  | "user"="root", "ip"="*"        |
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+

mysql> alter resource_queue test_queue4 set resource () set matching policy ();
Query OK, 0 rows affected (0.01 sec)

mysql> show all resource_queue;
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+
| QueueId | QueueName   | PendingJobs | RunningJobs | QueueConfig                                  | MatchingPolicy                 |
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+
|   74033 | test_queue4 |           0 |           0 | "max_concurrency"="1", "max_queue_size"="10" | "user"="root", "ip"="*"        |
|   74023 | test_queue1 |           2 |           1 | "max_concurrency"="1", "max_queue_size"="10" | "user"="ro*", "ip"="127.0.0.*" |
|   74026 | test_queue2 |           0 |           1 | "max_concurrency"="1", "max_queue_size"="0"  | "user"="root", "ip"="*"        |
|   74031 | test_queue3 |           0 |           1 | "max_concurrency"="2", "max_queue_size"="5"  | "user"="root", "ip"="*"        |
+---------+-------------+-------------+-------------+----------------------------------------------+--------------------------------+
```

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.91 seconds
 load time: 432 seconds
 storage size: 17172634117 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221120162119_clickbench_pr_49433.html

File: fe/fe-core/src/main/java/org/apache/doris/job/ResourceQueue.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import java.util.concurrent.atomic.AtomicLong;
+
+public class ResourceQueue {
+    private static final AtomicLong QUEUE_ID = new AtomicLong(0);
+
+    private final long id;
+    private final String name;
+    private final ResourceQueueConfig config;
+    private final MatchingPolicy policy;
+    private final AsyncJobManager jobMgr;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/job/ResourceQueue.java:37]: no need to call method of `jobMgr` in this class

File: fe/fe-core/src/main/java/org/apache/doris/job/ResourceQueue.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+    public int numRunningJobs() {
+        return runningJobs.size();
+    }
+
+    public long queueId() {
+        return id;
+    }
+
+    public void run() {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/job/ResourceQueue.java:123]: I think we should use one thread to schedule jobs in all queue.

File: fe/fe-core/src/main/java/org/apache/doris/job/AsyncJob.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+import org.apache.doris.analysis.UserIdentity;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+public abstract class AsyncJob {
+
+    public enum JobStatus {
+        INIT, PENDING, RUNNING, FINISHED, CANCELED
```
[the remark this answers is not in the document]
Comment (AshinGau) [reply fe/fe-core/src/main/java/org/apache/doris/job/AsyncJob.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/job/AsyncJob.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    private static final AtomicLong JOB_ID = new AtomicLong(0);
+    public static final String RESOURCE_QUEUE = "resource_queue";
+
+    // jobId
+    protected final long id;
+    protected final String label;
```
[the remark this answers is not in the document]
Comment (AshinGau) [reply fe/fe-core/src/main/java/org/apache/doris/job/AsyncJob.java:36]: If `AsyncJob`'s label is not set,  the label of `AsyncJob` will be set as the label of wrapped sql-statement.

File: fe/fe-core/src/main/java/org/apache/doris/job/ResourceQueue.java (hunk 2 above)
Comment (AshinGau) [reply fe/fe-core/src/main/java/org/apache/doris/job/ResourceQueue.java:237]: done

File: fe/fe-core/src/main/java/org/apache/doris/job/ResourceQueue.java (hunk 1 above)
Comment (AshinGau) [reply fe/fe-core/src/main/java/org/apache/doris/job/ResourceQueue.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/job/AsyncJobManager.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import org.apache.doris.common.UserException;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class AsyncJobManager {
+    private static volatile AsyncJobManager INSTANCE = null;
+
+    public static AsyncJobManager get() {
```
[the remark this answers is not in the document]
Comment (AshinGau) [reply fe/fe-core/src/main/java/org/apache/doris/job/AsyncJobManager.java:29]: I will create `AsyncJobManager` in `Env` when implementing running sql job.

**Pull request closed without merging**